applicationset/services/github_metrics.go

```diff
@@ -0,0 +1,244 @@
+		RateLimitLimit:     NewGitHubAPIRateLimitLimit(),
+		RateLimitReset:     NewGitHubAPIRateLimitReset(),
+		RateLimitUsed:      NewGitHubAPIRateLimitUsed(),
+	}
+}
+
+// Factory functions for each metric vector
+func NewGitHubAPIRequestTotal() *prometheus.CounterVec {
+	return prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: githubAPIRequestTotalMetricName,
+			Help: "Total number of GitHub API requests",
+		},
+		[]string{"method", "endpoint", "status", "appset_namespace", "appset_name"},
+	)
+}
+
+func NewGitHubAPIRequestDuration() *prometheus.HistogramVec {
+	return prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name:    githubAPIRequestDurationMetricName,
+			Help:    "GitHub API request duration in seconds",
+			Buckets: prometheus.DefBuckets,
+		},
+		[]string{"method", "endpoint", "appset_namespace", "appset_name"},
+	)
+}
+
+func NewGitHubAPIRateLimitRemaining() *prometheus.GaugeVec {
+	return prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: githubAPIRateLimitRemainingMetricName,
+			Help: "The number of requests remaining in the current rate limit window",
+		},
+		[]string{"endpoint", "appset_namespace", "appset_name", "resource", "credential_type"},
+	)
+}
+
+func NewGitHubAPIRateLimitLimit() *prometheus.GaugeVec {
+	return prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: githubAPIRateLimitLimitMetricName,
+			Help: "The maximum number of requests that you can make per hour",
+		},
+		[]string{"endpoint", "appset_namespace", "appset_name", "resource", "credential_type"},
+	)
+}
+
+func NewGitHubAPIRateLimitReset() *prometheus.GaugeVec {
+	return prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: githubAPIRateLimitResetMetricName,
+			Help: "The time left till the current rate limit window resets, in seconds",
+		},
+		[]string{"endpoint", "appset_namespace", "appset_name", "resource", "credential_type"},
+	)
+}
+
+func NewGitHubAPIRateLimitUsed() *prometheus.GaugeVec {
+	return prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: githubAPIRateLimitUsedMetricName,
+			Help: "The number of requests used in the current rate limit window",
+		},
+		[]string{"endpoint", "appset_namespace", "appset_name", "resource", "credential_type"},
+	)
+}
+
+// Global metrics (registered with the default registry)
+var globalGitHubMetrics = NewGitHubMetrics()
+
+func init() {
+	log.Debug("Registering GitHub API AppSet metrics")
+	metrics.Registry.MustRegister(globalGitHubMetrics.RequestTotal)
+	metrics.Registry.MustRegister(globalGitHubMetrics.RequestDuration)
+	metrics.Registry.MustRegister(globalGitHubMetrics.RateLimitRemaining)
+	metrics.Registry.MustRegister(globalGitHubMetrics.RateLimitLimit)
+	metrics.Registry.MustRegister(globalGitHubMetrics.RateLimitReset)
+	metrics.Registry.MustRegister(globalGitHubMetrics.RateLimitUsed)
+}
+
+type MetricsContext struct {
+	AppSetNamespace string
+	AppSetName      string
+	CredentialType  string // e.g., "github_app", "personal_token", "oauth_token"
```

Comment:
Refactor: Should be a type with predefined const values.

Or removed if we dont know how the credential type will be used. If someone uses multiple github apps, it will not be possible for them to know which one the metrics relate to by using this label.